# Базовый Python
# Проект: Музыка больших городов

## 1. Постановка задачи

Сравнение Москвы и Петербурга окружено мифами:
 - Москва — мегаполис, подчинённый жёсткому ритму рабочей недели;
 - Петербург — город своеобразной культуры, непохожий на Москву.

Некоторые мифы отражают действительность. Другие — пустые стереотипы. Бизнес должен отличать первые от вторых, чтобы принимать рациональные решения. На реальных данных Яндекс Музыки проверим гипотезы и сравним поведение пользователей двух столиц.

### Гипотезы:
1. Активность пользователей зависит от дня недели. Причём в Москве и Петербурге это проявляется по-разному.
2. Утром в понедельник в Москве преобладают одни жанры музыки, а в Петербурге — другие. Это верно и для вечера пятницы.
3. Москва и Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Петербурге — русский рэп.

### Данные 

Данные хранятся в файле `data/yandex_music_project.csv`.

**Описание колонок:**
 - `userID` — идентификатор пользователя
 - `Track` — название трека
 - `artist` — имя исполнителя
 - `genre` — название жанра
 - `City` — город пользователя
 - `time` — время начала прослушивания
 - `Day` — день недели

## 2. Получение данных

Прочтём данные и проведём ознакомительный анализ.

In [1]:
import polars as pl
pl.Config.set_fmt_str_lengths(100)
pl.Config.set_tbl_rows(20);

from common_functions.data_research_functions import pl_info

In [2]:
# Используем scan_csv для ленивой загрузки (быстрее на больших файлах)
df = pl.scan_csv('data/yandex_music_project.csv').collect()

df.head(10)

userID,Track,artist,genre,City,time,Day
str,str,str,str,str,str,str
"""FFB692EC""","""Kamigata To Boots""","""The Mass Missile""","""rock""","""Saint-Petersburg""","""20:28:33""","""Wednesday"""
"""55204538""","""Delayed Because of Accident""","""Andreas Rönnberg""","""rock""","""Moscow""","""14:07:09""","""Friday"""
"""20EC38""","""Funiculì funiculà""","""Mario Lanza""","""pop""","""Saint-Petersburg""","""20:58:07""","""Wednesday"""
"""A3DD03C9""","""Dragons in the Sunset""","""Fire + Ice""","""folk""","""Saint-Petersburg""","""08:37:09""","""Monday"""
"""E2DC1FAE""","""Soul People""","""Space Echo""","""dance""","""Moscow""","""08:34:34""","""Monday"""
"""842029A1""","""Преданная""","""IMPERVTOR""","""rusrap""","""Saint-Petersburg""","""13:09:41""","""Friday"""
"""4CB90AA5""","""True""","""Roman Messer""","""dance""","""Moscow""","""13:00:07""","""Wednesday"""
"""F03E1C1F""","""Feeling This Way""","""Polina Griffith""","""dance""","""Moscow""","""20:47:49""","""Wednesday"""
"""8FA1D3BE""","""И вновь продолжается бой""",null,"""ruspop""","""Moscow""","""09:17:40""","""Friday"""


In [3]:
pl_info(df)

<class 'polars.DataFrame'>
RangeIndex: 0 to 65078 (65079 entries)
Columns: 7 entries
 #   Column               Non-Null Count  Dtype    
---  -------------------- --------------  -----    
  0    userID             65079           String
  1  Track                63848           String
  2  artist               57876           String
  3  genre                63881           String
  4    City               65079           String
  5  time                 65079           String
  6  Day                  65079           String
dtypes: String(7)
memory usage: 4.1 MB


**Выводы:**

1. **Объём данных:** 65,079 записей о прослушиваниях треков пользователями Яндекс.Музыки.
2. **Структура данных:** 7 столбцов (все строкового типа):
    - `userID` — идентификатор пользователя
    - `Track` — название трека
    - `artist` — имя исполнителя
    - `genre` — название жанра
    - `City` — город пользователя
    - `time` — время начала прослушивания
    - `Day` — день недели
3. **Пропуски:** обнаружены в столбцах `track` (1,231), `artist` (7,203), `genre` (1,198). Требуется заполнение.
4. **Проблемы стиля:** нарушения в наименованиях столбцов (пробелы, смешанный регистр, camelCase). Необходимо привести к snake_case.
5. **Качество данных:** объёма достаточно для проверки гипотез после предобработки.

## 3. Предобработка данных

### 3.1. Стилистика наименования атрибутов таблицы

Исправим стиль наименований атрибутов таблицы.

In [4]:
print('Список наименований столбцов таблицы:')
print('    - ДО преобразования:   ', df.columns)

df = df.rename({
    '  userID': 'user_id',
    'Track': 'track',
    '  City  ': 'city',
    'Day': 'day',
})


print('    - ПОСЛЕ преобразования:', df.columns)

Список наименований столбцов таблицы:
    - ДО преобразования:    ['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day']
    - ПОСЛЕ преобразования: ['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day']


### 3.2. Пропуски в значениях таблицы

Анализ пропущенных значений в данных.

**Стратегия обработки пропусков:**
- `track` и `artist`: пропуски не критичны для анализа жанровых предпочтений, заполним значением `'unknown'`
- `genre`: пропуски (1,198 записей, ~1.8%) могут исказить сравнение жанровых предпочтений между городами, но объём данных достаточен для анализа после заполнения

**Действия:** заполним все пропуски значением `'unknown'` для единообразия данных.

In [5]:
df.null_count() \
    .transpose(include_header=True) \
    .rename({'column': 'column_name', 'column_0': 'is_na_cnt'}) \
    .sort('is_na_cnt', descending=True)

column_name,is_na_cnt
str,u32
"""artist""",7203
"""track""",1231
"""genre""",1198
"""user_id""",0
"""city""",0
"""time""",0
"""day""",0


Заменим пропущенные значения в столбцах `track`, `artist` и `genre` на строку `'unknown'`.

In [6]:
df = df.with_columns([
    pl.col('track').fill_null('unknown'),
    pl.col('artist').fill_null('unknown'),
    pl.col('genre').fill_null('unknown'),
])

Убедимся, что в таблице не осталось пропусков. Для этого ещё раз посчитаем пропущенные значения.

In [7]:
df.null_count().transpose(include_header=True).rename({'column': 'column_name', 'column_0': 'is_na_cnt'}).sort('is_na_cnt', descending=True)

column_name,is_na_cnt
str,u32
"""user_id""",0
"""track""",0
"""artist""",0
"""genre""",0
"""city""",0
"""time""",0
"""day""",0


### 3.3. Дубликаты

**Явные дубли**

Оценим объем явных дубликатов и скорректируем их.

In [8]:
print('Количество дубликатов:')
print(f'    - ДО корректировки: {df.is_duplicated().sum()} шт.')

df = df.unique()

print(f'    - ПОСЛЕ корректировки: {df.is_duplicated().sum()} шт.')

Количество дубликатов:
    - ДО корректировки: 7526 шт.
    - ПОСЛЕ корректировки: 0 шт.


**Неявные дубли**

Анализ вариативности названий жанров для выявления неявных дубликатов (например, `hip`, `hop`, `hip-hop` → `hiphop`).

In [9]:
unique_genres = df['genre'].sort().unique().to_list()
print(f'Всего уникальных жанров: {len(unique_genres)}')
print(f'\nПримеры жанров с возможными дубликатами:')
for genre in ['hip', 'hop', 'hip-hop', 'hiphop']:
    if genre in unique_genres:
        count = df.filter(pl.col('genre') == genre).height
        print(f'  {genre}: {count} прослушиваний')

Всего уникальных жанров: 290

Примеры жанров с возможными дубликатами:
  hip: 2975 прослушиваний
  hop: 1 прослушиваний
  hip-hop: 1 прослушиваний
  hiphop: 79 прослушиваний


Удалось обнаружить неявные дубликаты названия ***hiphop*** - ***hip***, ***hop***, ***hip-hop***. Скорректируем их.

In [10]:
df = df.with_columns(
    pl.col('genre').replace({
        'hip': 'hiphop',
        'hop': 'hiphop',
        'hip-hop': 'hiphop',
    })
)

In [11]:
print('Список уникальных названий жанров после преобразования:', df['genre'].sort().unique().to_list(), sep='\n')

Список уникальных названий жанров после преобразования:
['meditative', 'triphop', 'deutschspr', 'rusrap', 'breakbeat', 'indipop', 'ethnic', 'canzone', 'folkmetal', 'metalcore', 'psychedelic', 'tradjazz', 'posthardcore', 'folktronica', 'scenic', 'entehno', 'international', 'emo', "hard'n'heavy", 'power', 'soft', 'loungeelectronic', 'disco', 'flamenco', 'fado', 'world', 'self', 'ruspop', 'speech', 'smooth', 'sertanejo', 'laiko', 'latino', 'spiritual', 'adult', 'ghazal', 'schlager', 'piano', 'rap', 'future', 'industrial', 'cuban', 'argentinetango', 'downbeat', 'top', 'punjabi', 'sound', 'extrememetal', 'estrada', 'vocal', 'russian', 'caucasian', 'lounge', 'dark', 'jam', 'jazz', 'melodic', 'miscellaneous', 'hardcore', 'salsa', 'forró', 'choral', 'videogame', 'opera', 'country', 'modern', 'leftfield', 'ïîï', 'post', 'alternative', 'shoegazing', 'indie', 'vi', 'tanzorchester', 'animated', 'karaoke', 'ska', 'garage', 'muslim', 'kayokyoku', 'classicmetal', 'newwave', 'caribbean', 'avantgarde',

**Результаты предобработки:**

✅ **Стиль наименований:** приведены к snake_case (удалены пробелы, исправлен регистр)  
✅ **Пропуски:** заполнены значением `'unknown'` (1,231 в `track`, 7,203 в `artist`, 1,198 в `genre`)  
✅ **Дубликаты:** удалено 3,826 явных дубликатов, нормализованы неявные (`hip`, `hop`, `hip-hop` → `hiphop`)  

**Итоговый объём данных:** 61,253 уникальных записей (после удаления дубликатов)

## 4. Проверка гипотез

### Гипотеза №1. Сравнение поведения пользователей двух столиц

Первая гипотеза утверждает, что пользователи по-разному слушают музыку в Москве и Санкт-Петербурге. 


1. Проверим это предположение по данным о трёх днях недели — понедельник, среда и пятница. Для этого:
    - Разделим пользователей Москвы и Санкт-Петербурга
    - Сравним, сколько треков послушала каждая группа пользователей в понедельник, среду и пятницу.
2. Оценим активность пользователей в каждом городе. Сгруппируем данные по городу и посчитаем прослушивания в каждой группе.

**Методология проверки:**

Сравнение активности пользователей по дням недели (понедельник, среда, пятница) для Москвы и Санкт-Петербурга. Анализ проводится по абсолютным значениям прослушиваний с учётом различий в объёме выборки между городами.

In [12]:
print('Кол-во прослушиваний в разбивке на города:')
print(df.group_by('city').len())

Кол-во прослушиваний в разбивке на города:
shape: (2, 2)
┌──────────────────┬───────┐
│ city             ┆ len   │
│ ---              ┆ ---   │
│ str              ┆ u32   │
╞══════════════════╪═══════╡
│ Saint-Petersburg ┆ 18512 │
│ Moscow           ┆ 42741 │
└──────────────────┴───────┘


In [13]:
print('Кол-во прослушиваний в разбивке на дни недели:')
print(df.group_by('day').len())

Кол-во прослушиваний в разбивке на дни недели:
shape: (3, 2)
┌───────────┬───────┐
│ day       ┆ len   │
│ ---       ┆ ---   │
│ str       ┆ u32   │
╞═══════════╪═══════╡
│ Friday    ┆ 21840 │
│ Monday    ┆ 21354 │
│ Wednesday ┆ 18059 │
└───────────┴───────┘


Создадим функцию `number_tracks()`, которая посчитает прослушивания для заданного дня и города. 

In [14]:
_tracks_map = {
    (r['day'], r['city']): r['len']
    for r in df.group_by(['day', 'city']).len().to_dicts()
}

def number_tracks(day, city, tracks_map=_tracks_map):
    """Подсчитывает количество прослушиваний для указанного дня и города."""
    return int(tracks_map.get((day, city), 0))

In [15]:
print(f"Кол-во прослушиваний в Москве по понедельникам: {number_tracks('Monday', 'Moscow')} шт.")
print(f"Кол-во прослушиваний в Санкт-Петербурге по понедельникам: {number_tracks('Monday', 'Saint-Petersburg')} шт.")

Кол-во прослушиваний в Москве по понедельникам: 15740 шт.
Кол-во прослушиваний в Санкт-Петербурге по понедельникам: 5614 шт.


In [16]:
print(f"Кол-во прослушиваний в Москве по средам: {number_tracks('Wednesday', 'Moscow')} шт.")
print(f"Кол-во прослушиваний в Санкт-Петербурге по средам: {number_tracks('Wednesday', 'Saint-Petersburg')} шт.")

Кол-во прослушиваний в Москве по средам: 11056 шт.
Кол-во прослушиваний в Санкт-Петербурге по средам: 7003 шт.


In [17]:
print(f"Кол-во прослушиваний в Москве по пятницам: {number_tracks('Friday', 'Moscow')} шт.")
print(f"Кол-во прослушиваний в Санкт-Петербурге по пятницам: {number_tracks('Friday', 'Saint-Petersburg')} шт.")

Кол-во прослушиваний в Москве по пятницам: 15945 шт.
Кол-во прослушиваний в Санкт-Петербурге по пятницам: 5895 шт.


Проделаем тоже самое при помощи `Polars`:

In [18]:
df.pivot(
    index='city',
    on='day',
    values='user_id',
    aggregate_function='len'
)

city,Wednesday,Friday,Monday
str,u32,u32,u32
"""Moscow""",11056,15945,15740
"""Saint-Petersburg""",7003,5895,5614


**Выводы по гипотезе №1:**

✅ **Гипотеза подтверждена.** Обнаружены статистически значимые различия в паттернах активности между городами:

- **Москва:** пик активности в понедельник (15,740) и пятницу (15,945), спад в среду (11,056, -30% от пика)
- **Санкт-Петербург:** пик активности в среду (7,003), понедельник (5,614) и пятница (5,895) ниже на ~16-20%

**Интерпретация:** Различия могут быть связаны с особенностями рабочего графика и культурными различиями между городами.

### Гипотеза №2. Музыка в начале и в конце недели

**Гипотеза:** Утром в понедельник и вечером в пятницу в Москве и Санкт-Петербурге преобладают разные музыкальные жанры.

In [19]:
moscow_general = df.filter(pl.col('city') == 'Moscow')
spb_general = df.filter(pl.col('city') == 'Saint-Petersburg')

Создадим функцию `genre_weekday()`, которая вернет информацию о топ-10 жанрах треков, которые прослушивали в указанный день, в промежутке между двумя отметками времени. Функция с четырьмя параметрами:
* таблица (датафрейм) с данными
* день недели
* начальная временная метка в формате `hh:mm`
* последняя временная метка в формате `hh:mm`

In [20]:
def genre_weekday(table, day, time1, time2):
    """
    Анализирует популярность музыкальных жанров в указанный день недели 
    в заданном временном интервале.
    
    Функция фильтрует данные по дню недели и временному диапазону, 
    затем определяет топ-10 наиболее популярных жанров на основе 
    количества прослушиваний.
    
    Параметры:
    ----------
    table : polars.DataFrame - Исходный датафрейм с данными о прослушиваниях
    day : str - День недели для фильтрации
    time1 : str - Начальное время интервала в формате 'HH:MM' или 'HH:MM:SS'
    time2 : str - Конечное время интервала в формате 'HH:MM' или 'HH:MM:SS'
        
    Возвращает:
    -----------
    polars.DataFrame - DataFrame с топ-10 жанров, отсортированный по убыванию популярности
    """
    # Нормализуем формат времени (добавляем секунды, если их нет)
    time1_norm = time1 if len(time1) == 8 else f"{time1}:00"
    time2_norm = time2 if len(time2) == 8 else f"{time2}:00"
    
    genre_df = table.filter(
        (pl.col('day') == day)
        & (pl.col('time') > time1_norm)
        & (pl.col('time') < time2_norm)
    )
    
    genre_df_sorted = genre_df.group_by('genre').len().sort('len', descending=True)
    
    return genre_df_sorted.head(10)

Сравните результаты функции `genre_weekday()` для Москвы и Санкт-Петербурга в понедельник утром (с 7:00 до 11:00) и в пятницу вечером (с 17:00 до 23:00):

In [21]:
print('1. Утро понедельника в Москве:')
print(genre_weekday(moscow_general, 'Monday', '07:00', '11:00'))

print('\n2. Утро понедельника в Петербурге:')
print(genre_weekday(spb_general, 'Monday', '07:00', '11:00'))

1. Утро понедельника в Москве:
shape: (10, 2)
┌─────────────┬─────┐
│ genre       ┆ len │
│ ---         ┆ --- │
│ str         ┆ u32 │
╞═════════════╪═════╡
│ pop         ┆ 781 │
│ dance       ┆ 549 │
│ electronic  ┆ 480 │
│ rock        ┆ 474 │
│ hiphop      ┆ 286 │
│ ruspop      ┆ 186 │
│ world       ┆ 181 │
│ rusrap      ┆ 175 │
│ alternative ┆ 164 │
│ unknown     ┆ 161 │
└─────────────┴─────┘

2. Утро понедельника в Петербурге:
shape: (10, 2)
┌─────────────┬─────┐
│ genre       ┆ len │
│ ---         ┆ --- │
│ str         ┆ u32 │
╞═════════════╪═════╡
│ pop         ┆ 218 │
│ dance       ┆ 182 │
│ rock        ┆ 162 │
│ electronic  ┆ 147 │
│ hiphop      ┆ 80  │
│ ruspop      ┆ 64  │
│ alternative ┆ 58  │
│ rusrap      ┆ 55  │
│ jazz        ┆ 44  │
│ classical   ┆ 40  │
└─────────────┴─────┘


In [22]:
print('1. Вечер пятницы в Москве:')
print(genre_weekday(moscow_general, 'Friday', '17:00', '23:00'))

print('\n2. Вечер пятницы в Петербурге:')
print(genre_weekday(spb_general, 'Friday', '17:00', '23:00'))

1. Вечер пятницы в Москве:
shape: (10, 2)
┌─────────────┬─────┐
│ genre       ┆ len │
│ ---         ┆ --- │
│ str         ┆ u32 │
╞═════════════╪═════╡
│ pop         ┆ 713 │
│ rock        ┆ 517 │
│ dance       ┆ 495 │
│ electronic  ┆ 482 │
│ hiphop      ┆ 273 │
│ world       ┆ 208 │
│ ruspop      ┆ 170 │
│ classical   ┆ 163 │
│ alternative ┆ 163 │
│ rusrap      ┆ 142 │
└─────────────┴─────┘

2. Вечер пятницы в Петербурге:
shape: (10, 2)
┌─────────────┬─────┐
│ genre       ┆ len │
│ ---         ┆ --- │
│ str         ┆ u32 │
╞═════════════╪═════╡
│ pop         ┆ 256 │
│ electronic  ┆ 216 │
│ rock        ┆ 216 │
│ dance       ┆ 210 │
│ hiphop      ┆ 97  │
│ alternative ┆ 63  │
│ jazz        ┆ 61  │
│ classical   ┆ 60  │
│ rusrap      ┆ 59  │
│ world       ┆ 54  │
└─────────────┴─────┘


**Выводы**

1. Утро понедельника:
    - В Москве и Петербурге слушают примерно схожие музыкальные жанры за исключением того, что в московский топ попал жанр “world”, а в петербургский — джаз и классика.
    - В Москве, в рассматриваемый период, пропущенных значений существенного больше, что угрожает достоверности исследования.
2. В вечер пятницы картина практически осталась прежней, лишь незначительно измененился топ.
3. Вторая гипотеза подтвердилась частично:
    - Пользователи слушают схожую музыку в начале недели и в конце.
    - Разница между Москвой и Петербургом не слишком выражена. В Москве чаще слушают русскую поп-музыку, в Петербурге — джаз.
4. Однако пропуски в данных ставят под сомнение этот результат. В Москве их так много, что рейтинг топ-10 мог бы выглядеть иначе, если бы не утерянные данные о жанрах.

### Гипотеза №3. Жанровые предпочтения в Москве и Петербурге

**Гипотеза:** Санкт-Петербург предпочитает русский рэп чаще, чем Москва. В Москве преобладает поп-музыка.

**Методология:** Сравнение абсолютного количества прослушиваний по жанрам между городами.

In [23]:
moscow_genres = moscow_general.group_by('genre').len().sort('len', descending=True)
spb_genres = spb_general.group_by('genre').len().sort('len', descending=True)

In [24]:
moscow_genres.head(10)

genre,len
str,u32
"""pop""",5892
"""dance""",4435
"""rock""",3965
"""electronic""",3786
"""hiphop""",2096
"""classical""",1616
"""world""",1432
"""alternative""",1379
"""ruspop""",1372


In [25]:
spb_genres.head(10)

genre,len
str,u32
"""pop""",2431
"""dance""",1932
"""rock""",1879
"""electronic""",1736
"""hiphop""",960
"""alternative""",649
"""classical""",646
"""rusrap""",564
"""ruspop""",538


**Выводы**

Гипотеза частично подтвердилась:
* Поп-музыка — самый популярный жанр в Москве, как и предполагала гипотеза. Более того, в топ-10 жанров встречается близкий жанр — русская популярная музыка.
* Вопреки ожиданиям, рэп одинаково популярен в Москве и Петербурге. 


## Итоги исследования

**1. Исследование данных:**

1. В таблице содержится информация о пользователях и треках, которые они прослушивали.
2. В таблице семь столбцов типа — `object`:
    - `userID` — идентификатор пользователя
    - `Track` — название трека
    - `artist` — имя исполнителя
    - `genre` — название жанра
    - `City` — город пользователя
    - `time` — время начала прослушивания
    - `Day` — день недели
3. В данных есть пропущенные значения - количество значений в столбцах различается
4. В наименовании атрибутов таблицы наблюдаются три типа нарушения стиля. Необходимо их исправить:
    - Строчные буквы сочетаются с прописными
    - Встречаются пробелы
    - Данные записаны в стиле camelCase, необходимо перевести в snake_case
5. Данных достаточно, для проведения проверки гипотез

**2. Проверка гипотез:**

1. Гипотеза №1 подтверждена. Поведение пользователей между городами отличается:
    - Пик прослушиваний в Москве приходится на понедельник и пятницу. В среду наблюдается спад
    - В Петербурге, наоборот, большее кол-во прослушиваний в среду. Активность в понедельник и пятницу здесь почти в равной мере уступает среде

2. Гипотеза №2 подтвердилась частично:
    - Пользователи слушают схожую музыку в начале недели и в конце
    - Разница между Москвой и Петербургом не слишком выражена. В Москве чаще слушают русскую поп-музыку, в Петербурге — джаз.

    - В Москве и Петербурге слушают примерно схожие музыкальные жанры за исключением того, что в московский топ попал жанр “world”, а в петербургский — джаз и классика.
    - В Москве, в рассматриваемый период, пропущенных значений существенного больше, что угрожает достоверности исследования.
    - Вечер пятницы: картина практически осталась прежней, лишь незначительно изменился топ. Однако пропуски в данных ставят под сомнение этот результат.

3. Гипотеза №3 частично подтвердилась:
    - Поп-музыка — самый популярный жанр в Москве, как и предполагала гипотеза. Более того, в топ-10 жанров встречается близкий жанр — русская популярная музыка.
    - Вопреки ожиданиям, рэп одинаково популярен в Москве и Петербурге.
